### PR TITLE
[ISSUE #690] Support batch msgs in dledger mode

### PR DESCRIPTION
**Pull request by TerrellChen**
## What is the purpose of the change

https://github.com/apache/rocketmq/issues/690

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by RongtongJin** [comment]
Good. job! I will review the code ASAP.

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java [hunk 1]
```diff
@@ -507,7 +513,122 @@ public PutMessageResult putMessage(final MessageExtBrokerInner msg) {
...
+                    msgIdBuilder.append(',').append(msgId);
+                } else {
+                    msgIdBuilder.append(msgId);
+                }
+                msgNum++;
+            }
+            elapsedTimeInLock = this.defaultMessageStore.getSystemClock().now() - beginTimeInDledgerLock;
+            appendResult = new AppendMessageResult(AppendMessageStatus.PUT_OK, wroteOffset, encodeResult.totalMsgLen,
+                    msgIdBuilder.toString(), System.currentTimeMillis(), queueOffset, elapsedTimeInLock);
+            DLedgerCommitLog.this.topicQueueTable.put(encodeResult.queueOffsetKey, queueOffset + msgNum);
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java:585]: When sending batch messages, worteOffset is the offset of the first message, but this is the offset of the last message, so the unit test fails.

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java [hunk 2]
```diff
@@ -507,7 +513,122 @@ public PutMessageResult putMessage(final MessageExtBrokerInner msg) {
...
+            for (long pos : dledgerFuture.getPositions()) {
+                wroteOffset = pos + DLedgerEntry.BODY_OFFSET;
+                String msgId = MessageDecoder.createMessageId(buffer, messageExtBatch.getStoreHostBytes(), wroteOffset);
+                if (msgIdBuilder.length() > 0) {
+                    msgIdBuilder.append(',').append(msgId);
+                } else {
+                    msgIdBuilder.append(msgId);
+                }
+                msgNum++;
+            }
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java:581]: Write code as this:
```java
boolean isFirstOffset = true;
            for (long pos : dledgerFuture.getPositions()) {
                if(isFirstOffset) {
                    wroteOffset = pos + DLedgerEntry.BODY_OFFSET;
                    isFirstOffset = false;
                }
                String msgId = MessageDecoder.createMessageId(buffer, messageExtBatch.getStoreHostBytes(), wroteOffset);
                if (msgIdBuilder.length() > 0) {
                    msgIdBuilder.append(',').append(msgId);
                } else {
                    msgIdBuilder.append(msgId);
                }
                msgNum++;
            }
```
Or do you have a better solution？

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java [hunk 3]
```diff
@@ -507,7 +513,122 @@ public PutMessageResult putMessage(final MessageExtBrokerInner msg) {
...
+                return new PutMessageResult(PutMessageStatus.MESSAGE_ILLEGAL, new AppendMessageResult(encodeResult
+                        .status));
+            }
+            BatchAppendEntryRequest request = new BatchAppendEntryRequest();
+            request.setGroup(dLedgerConfig.getGroup());
+            request.setRemoteId(dLedgerServer.getMemberState().getSelfId());
+            request.setBatchMsgs(encodeResult.batchData);
+            dledgerFuture = (BatchAppendFuture<AppendEntryResponse>) dLedgerServer.handleAppend(request);
+            if (dledgerFuture.getPos() == -1) {
+                log.warn("[DEBUG_CTR] handleAppend return false due to error code {}", dledgerFuture.get().getCode());
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java:564]: Removing `[DEBUG_CTR]` will be better.

**Comment by RongtongJin** [comment]
@TerrellChen  I have released the dledger version to 0.2.2 which includes the bug you fixed. You can upgrade the dledger version when you have time.

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java (hunk 1 above)
Comment (TerrellChen) [reply store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java:585]: Good job!

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java (hunk 2 above)
Comment (TerrellChen) [reply store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java:581]: LGTM!

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java (hunk 3 above)
Comment (TerrellChen) [reply store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java:564]: My fault!

**Comment by TerrellChen** [comment]
@RongtongJin Thank you for all your assistance! 
I've made a commit by your comment. Please check it again.

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java [hunk 4]
```diff
@@ -609,7 +735,122 @@ public PutMessageResult putMessages(final MessageExtBatch messageExtBatch) {
...
+                    isFirstOffset = false;
+                }
+                String msgId = MessageDecoder.createMessageId(buffer, messageExtBatch.getStoreHostBytes(), wroteOffset);
+                if (msgIdBuilder.length() > 0) {
+                    msgIdBuilder.append(',').append(msgId);
+                } else {
+                    msgIdBuilder.append(msgId);
+                }
+                msgNum++;
+            }
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java:807]: This is my fault. The code will cause the writeOffset of the following message not to be updated, so as to get the wrong msgId. We need to fix it and ensure that the writeOffset of the first message is returned to the user.

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java [hunk 5]
```diff
@@ -507,7 +513,127 @@ public PutMessageResult putMessage(final MessageExtBrokerInner msg) {
...
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java:585]: Same as below

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java (hunk 4 above)
Comment (TerrellChen) [reply store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java]: fixed

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java (hunk 5 above)
Comment (TerrellChen) [reply store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java]: fixed

**Review by RongtongJin — APPROVED** [review]

**Review by ShannonDing — APPROVED** [review]

**Pull request merged**